base_layer/core/src/proof_of_work/monero_rx.rs

```diff
@@ -103,16 +176,99 @@ fn monero_difficulty_calculation(header: &BlockHeader) -> Result<Difficulty, Mer
     }
 }
 
-fn create_input_blob(_data: &MoneroData) -> Result<String, MergeMineError> {
-    // Todo deserialize monero data to create string for  randomX vm
-    // Returning an error here so that difficulty can return 0 as this is not yet implemented.
-    Err(MergeMineError::HashingError)
+/// Appends merge mining hash to a Monero block and returns the Monero blocktemplate_blob
+pub fn append_merge_mining_tag(block: &MoneroBlock, hash: Hash) -> Result<String, MergeMineError> {
+    #[cfg(feature = "monero_merge_mining")]
+    {
+        let mut monero_block = block.clone();
+        let mm_tag = SubField::MergeMining(VarInt(0), hash);
+        monero_block.miner_tx.prefix.extra.0.push(mm_tag);
+        let serialized = serialize::<MoneroBlock>(&block);
+        Ok(hex::encode(&serialized).into())
+    }
+    #[cfg(not(feature = "monero_merge_mining"))]
+    {
+        Err(MergeMineError::HashingError)
+    }
+}
+
+/// Calculates the Monero blockhashing_blob
+pub fn create_input_blob(
+    header: &MoneroBlockHeader,
+    tx_count: &u16,
+    tx_hashes: &Vec<[u8; 32]>,
+) -> Result<String, MergeMineError>
+{
+    #[cfg(feature = "monero_merge_mining")]
+    {
+        let header = serialize::<MoneroBlockHeader>(header);
+        // Note count assumes the miner tx is included already
+        let mut count = serialize::<VarInt>(&VarInt(tx_count.clone() as u64));
+        let mut hashes = Vec::new();
+        for item in tx_hashes {
+            hashes.push(Hash(from_slice(item.clone().as_bytes())));
+        }
+        let mut root = tree_hash(hashes);
+        let mut encode2 = header;
+        encode2.append(&mut root);
+        encode2.append(&mut count);
+        Ok(hex::encode(encode2).into())
+    }
+    #[cfg(not(feature = "monero_merge_mining"))]
+    {
+        Err(MergeMineError::HashingError)
+    }
+}
+
+/// Utility function to transform array to fixed array
+pub fn from_slice(bytes: &[u8]) -> [u8; 32] {
+    let mut array = [0; 32];
+    let bytes = &bytes[..array.len()]; // panics if not enough data
+    array.copy_from_slice(bytes);
+    array
+}
+
+/// Utility function to transform array of hash to fixed array of [u8; 32]
+pub fn from_hashes(hashes: &[Hash]) -> Vec<[u8; 32]> {
+    let mut result = Vec::new();
+    for item in hashes {
+        result.push(item.0);
+    }
+    result
+}
+
+fn verify_root(monero_data: &MoneroData) -> Result<(), MergeMineError> {
+    let mut hashes = Vec::new();
+    for item in &monero_data.transaction_hashes {
+        hashes.push(Hash(from_slice(item.to_vec().as_slice())));
+    }
+    let root = tree_hash(hashes);
+
+    if !(monero_data.transaction_root.to_vec() == root) {
+        return Err(MergeMineError::ValidationError);
```

Comment:
I think it will be useful to add a reason or message to distinguish this ValidationError from the one below. E.g. 
```suggestion
        return Err(MergeMineError::ValidationError("Transaction root did not match");
```